politeiawww/sessionstore.go

```diff
@@ -0,0 +1,184 @@
+package main
+
+import (
+	"encoding/base32"
+	"errors"
+	"net/http"
+	"strings"
+
+	"github.com/decred/politeia/politeiawww/user"
+	"github.com/google/uuid"
+	"github.com/gorilla/securecookie"
+	"github.com/gorilla/sessions"
+)
+
+// SessionStore stores sessions in the database.
+//
+// Please note: this is (by and large) a clone of goriall mux'
+// `sessions.FilesystemStore` with the save(), load() and erase()
+// methods adapted to use the database as the backing storage.
+type SessionStore struct {
```

Comment:
Makes sense. I will only be able to revisit this next week though. Fully slammed this week :-P

Reply:
No problem.

Reply:
Alright, I took a look at this and it turns out that `FilesystemStore` does not lend itself to being embedded.

I have 2 choices which are equally ugly:
  1. use `FilesystemStore.Save()` in which case we need to maintain 2 session stores, one in `FilesystemStore` and the other in the database which is tricky and complex
  2. **not** use `FilesystemStore.Save()` but use/keep the `SessionStore.Save()` func as is, in which case we gain very little from embedding `FilesystemStore` 

In light of this I would suggest to leave the code as is..

Reply:
Why doesn't `FilesystemStore` lend itself to being embedded?

Reply:
This pull request adds the capability to remove all sessions for a given user by calling `SessionsDeleteByUserId()`. Calling that function will remove all the user's sessions from the database but an embedded `FileSystemStore` instance will continue to return obsolete session objects since the "session" cookie has not been reset for the 1+ client connections.

IOW, it is just not very useful. We still need to look at the database (as the single source of truth) and cannot reuse `FilesystemStore.New()` as it stands.

Reply:
Gotcha.  Thanks for the explanation.  Let me talk this over with Marco before coming to a decision, but I think you're probably right that the best course of action is to leave the code as it.

Reply:
So, what's the verdict? Are we merging this or discarding it?